kythe/proto/pipeline.proto

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+syntax = "proto3";
+
+package kythe.proto.pipeline;
+
+option java_package = "com.google.devtools.kythe.proto";
+
+import "kythe/proto/schema.proto";
+import "kythe/proto/storage.proto";
+
+// A Node represents a Kythe graph node with all its known facts and edges.
+message Node {
```

Comment:
The term that we're using for this concept internally is "Source".  Calling it "Node" is confusing IMO because it's not a node per se, but effectively a union that can represent either a node and its facts, or a node and its associated edge data (although see my comment on `nodes.go` about that).  Any reason not to go with the existing internal name?

Please add documentation that specifies what constraints, if any, should be imposed on which sets of fields are expected to be simultaneously populated.  (The existing code in `nodes.go` implies that either `fact` and the `*kind` fields will be populated, or `edge` will be populated, but not both.)

Reply:
The Source proto is also external.

But no, it is Node in the Kythe graph.  A Kythe node is composed of all of its edges and its facts (not either).  This exactly the same as the Source proto without the special-cases and enums.  There's not a whole lot of constraints on this proto as it will be filtered and massaged throughout the pipeline.

Reply:
You're right, the `Source` proto is also external, I got briefly confused because it's in `internal.proto`.  :P

So if this is intended to fill the same ecological niche as `Source`, would it make sense to augment it to include the special cases and enums rather than adding a new message?

Thanks for the clarification on this proto's use.  Would it make sense to add an enum that would specify what type of use case a given message is satisfying?  (All the current uses look like they're either storing facts or edges but not both, but I understand that there will be other uses later.)  That would make validation possible.

Reply:
I'd rather make a clean break from the current system.  It removes the need for churn elsewhere.

The current use is storing both facts and edges.  The other "types" of uses are really just filtered views, not much to tag an enum on.